pupil_src/shared_modules/marker_tracker_3d/controller.py

```diff
@@ -0,0 +1,26 @@
+from marker_tracker_3d import optimization
+from marker_tracker_3d.basic_models import CameraModel, MarkerModel
+from marker_tracker_3d.camera_localizer import CameraLocalizer
+from marker_tracker_3d.marker_detector import MarkerDetector
+
+
+class Controller:
+    def __init__(self, storage, update_menu):
+        self.storage = storage
+        self.storage.camera_model = CameraModel(cameraMatrix=None, distCoeffs=None)
+        self.storage.marker_model = MarkerModel()
+
+        self.marker_detector = MarkerDetector(self.storage)
+        self.optimization_controller = optimization.Controller(
+            self.storage, update_menu
+        )
+        self.camera_localizer = CameraLocalizer(self.storage)
+
+    def recent_events(self, frame):
```

Comment:
I like how this functions gives a nice overview of what is happening overall. :+1: I think we can improve it even more!

I remember we have talked about how it is nicer to have functions that take arguments and return something. The function calls here are a good example I think. For example if I read `fetch_extrinsics` I would expect that something is returned, after all this function is fetching something. Currently I can not see where the fetched data is going. Similarly with `send_marker_data` I can not see what data is being send, and `update` does not tell me what exactly is being updated.

It would therefore be nice to rewrite this to something like this (I did not test this!):

Old:
```python
self.marker_detector.detect(frame)

self.optimization_controller.fetch_extrinsics()

self.camera_localizer.update()

self.optimization_controller.send_marker_data()
```
New
```python
self.storage.markers = self.marker_detector.detect(frame)

self.storage.marker_extrinsics, self.storage.marker_points_3d = (
    self.optimization_controller.fetch_extrinsics()
)

self.storage.camera_extrinsics = self.camera_localizer.update(
    self.storage.markers,
    self.storage.marker_extrinsics,
    self.storage.camera_extrinsics_previous,
)

self.optimization_controller.send_marker_data(
    self.storage.markers, self.storage.camera_extrinsics
)
```

With such a change it might not even be necessary anymore for all the individual controller objects to own `storage` anymore, since they get the data they need as arguments. I think this could be great for decoupling everything.